pkg/helm/urlwatcher.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helm
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"istio.io/pkg/log"
+)
+
+// poller is used to poll files from remote url at specific internal
+type poller struct {
+	// url is remote target url to poll from
+	url string
+	// minInterval is time intervals in minutes for polling
+	minInterval time.Duration
+	// existingHash records last sha value of polled files
+	existingHash string
+	// time ticker
+	ticker *time.Ticker
+}
+
+const (
+	// InstallationChartsFileName is sample Chart filename to fetch
+	InstallationChartsFileName = "istio-installer.tar.gz"
+	// InstallationShaFileName is Sha filename to verify
+	InstallationShaFileName = "istio-installer.tar.gz.sha256"
+	// VersionsFileName is version file's name
+	VersionsFileName = "VERSIONS.yaml"
+	// ChartsTempFilePrefix is temporary Files prefix
+	ChartsTempFilePrefix = "charts-"
+)
+
+// checkUpdate checks whether sha file being updated or not
+// fetch latest charts if yes
+func (p *poller) checkUpdate(uf *URLFetcher) error {
```

Comment:
originally I put that as part of poller, but later I make them more independent because I think poller can be used with some other logics as well, not necessarily the URLfetcher

Reply:
then, you could redesign the poller New function to take the new fetcher. otherwise you're forcing the user to maintain two separate objects that belong together.
also, unless you have a specific scenario in mind that looks likely it's best to follow YAGNI.

Reply:
done